enhancements/storage/storage-performant-security-policy.md

```diff
@@ -0,0 +1,202 @@
+
+We also want to allow Openshift admins to set `selinuxChangePolicy` for certain namespaces, so as they can opt-out of potentially breaking changes being introduced via `SELinuxMount` feature upcoming in future k8s releases.
+
+## Motivation
+
+### Motivation for fsgroupchangepolicy defaulting
+
+Currently it can take a long time for volume's permission to be changed to match pod's `fsGroup`. This is a well known problem
+in Kubernetes and Openshift. See - https://access.redhat.com/solutions/6221251 for more details.
+
+To solve this problem, we introduced `fsGroupChangePolicy` - https://github.com/kubernetes/kubernetes/blob/b5608aea94cfb54fea3a63e1d74235759d036c51/pkg/apis/core/types.go#L3977 , which
+can be configured in a `PodSpec` and be set to `OnRootMismatch`, which instructs kubelet to stop recursively changing permissions of a volume, if top level directory permissions match as expected by kubelet.
+
+This usually speeds up pod startup because kubelet does not have to recursively change permission of each file and directory in the volume. This mechanism relies on heuristic that gets applied when recursively changing permissions of a volume -  kubelet changes permissions of top level directory as a last step in the entire process, when it performs full recursive `chown` and `chmod` of the volume.
+
+Having said that, `fsGroupChangePolicy` still defaults to `Always`, which means kubelet will always recursively change permissions. The reason for defaulting to `Always` is - mostly backward compatibility and if users bring in volumes that were used outside kubernetes. In these cases `OnRootMismatch` may not be a safe default.
+
+But for most intent and purpose, it can be a reasonable default and hence this proposal allows admins to configure a namespace such that, all pods created in the namespace will default to `OnRootMismatch` policy if pod does not specify a policy of its own. If a pod specifies its own `fsGroupChangePolicy` that is still used as a default and will not be overridden.
+
+### Motivation for selinuxChangePolicy defaulting
+
+https://github.com/kubernetes/enhancements/issues/1710 is bringing changes which can be breaking for certain pods which use different selinux contexts for same volume between different pods or same pod.
+
+This will speed up time spent while recursively changing selinux policies of a volume.
+
+By allowing cluster-admins to set namespace default of `selinuxChangePolicy` we want to give control over how these changes are applied to Openshift clusters, without explicitly  changing pods themselves.
+
+### User Stories
+
+N/A
+
+### Goals
+
+- Allow Openshift admins to configure per-namespace default of `fsGroupChangePolicy`.
+- Allow Openshift admins to configure per-namespace default of `selinuxChangePolicy`.
+
+### Non-Goals
+
+* Change Openshift default globally.
```

Comment:
So this goes back to an earlier comment I had, are we sure we don't want to suggest the newer shinier options as the default for _new_ clusters?

It's not uncommon for us to do similar, but preserve the previous behaviour for clusters that upgrade

Reply:
It may not be sufficiently explained for selinux policy and I noticed a typo there, which can muddle the waters. Let me update it and add more wording.

Reply:
I have added more details for selinux policies here -  https://github.com/openshift/enhancements/pull/1804/commits/3a3eba791fad480a059aaa5bd07331a4efe36b80#diff-f35a977be72325492acea12227900f09b04aabea49b49aa78543cb12f3de8a4dR57